Userland/Utilities/cp.cpp

```diff
@@ -51,13 +51,22 @@ int main(int argc, char** argv)
     args_parser.add_positional_argument(destination, "Destination file path", "destination");
     args_parser.parse(argc, argv);
 
-    auto my_umask = umask(0);
-    umask(my_umask);
-
     for (auto& source : sources) {
-        bool ok = copy_file_or_directory(source, destination, recursion_allowed, link);
-        if (!ok)
+        auto rc = Core::File::copy_file_or_directory(
+            destination, source,
+            recursion_allowed ? Core::File::RecursionMode::Allowed : Core::File::RecursionMode::Disallowed,
+            link ? Core::File::LinkMode::Allowed : Core::File::LinkMode::Disallowed,
+            Core::File::AddDuplicateFileMarker::No);
+
+        if (rc.is_error()) {
+            // FIXME: This feels a bit hacky, but I am not sure what a better way would be
+            if (rc.error() == "Tried to copy directory with RecursionMode::Disallowed")
```

Comment:
Way too hacky indeed; either use `OSError` or if we want to provide even more info then let's define a richer error struct type and return that from `copy_file_or_directory`.

Reply:
Is the richer error type I've written a good fit in your opinion?